allennlp/data/lazy_dataset.py

```diff
@@ -0,0 +1,114 @@
+"""
+A :class:`~LazyDataset` is a subclass of :class:`~Dataset` that
+instead of storing a ``List`` of ``Instance`` s, stores a method
+for generating instances. The main intended use case is for datasets
+that are too large to load into memory, in which case the generator
+would just read the file from disk for each call to ``iterinstances``.
+"""
+import logging
+from collections import defaultdict
+from typing import Dict, Union, Callable, Iterator
+
+import torch
+import tqdm
+
+from allennlp.data.dataset import Dataset
+from allennlp.data.instance import Instance
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.common.checks import ConfigurationError
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class LazyDataset(Dataset):
+    """
+    A collection of :class:`~allennlp.data.instance.Instance` objects.
+    The ``Instances`` have ``Fields``, and the fields
+    could be in an indexed or unindexed state - the ``Dataset`` has methods around indexing the
+    data and converting the data into arrays.
+    """
+    def __init__(self, generator: Callable[[], Iterator[Instance]]) -> None:
+        """
+        A LazyDataset just takes a way of generating instances.
+        """
+        # Call superclass constructor with no instances, we'll override the methods that use them.
+        super().__init__([])
```

Comment:
I don't like this much - see my comment below about possibly removing this whole class and supporting laziness in `Dataset` by passing a flag.